### PR TITLE
Fix test caching on Windows/GNU

### PR DESCRIPTION
**Pull request by petrochenkov**
Addresses https://github.com/rust-lang/rust/issues/36385#issuecomment-277131231

Previously the sysroot directory was purged on every build and mingw startup objects were rebuilt unconditionally and always triggered test reruns.
Now mingw startup objects are built in the native directory and then copied into the sysroot directory. They are also rebuilt only when necessary, so test caching works.

### CODE REVIEW

File: src/bootstrap/compile.rs [hunk 1]
```diff
@@ -294,9 +298,7 @@ fn compiler_file(compiler: &Path, file: &str) -> PathBuf {
 }
 
 pub fn create_sysroot(build: &Build, compiler: &Compiler) {
-    let sysroot = build.sysroot(compiler);
-    let _ = fs::remove_dir_all(&sysroot);
```
Comment (alexcrichton) [inline src/bootstrap/compile.rs]: This is actually a somewhat crucial step as it cleans out any old artifacts to make sure we don't preserve anything by accident. Could the `*.o` files above be generated into something like `build/$target/rtstartup/foo.o` and then copied from there to the sysroot?

**Review by alexcrichton — COMMENTED** [review]
Looks great, thanks @petrochenkov!

File: src/bootstrap/compile.rs (hunk 1 above)
Comment (petrochenkov) [inline src/bootstrap/compile.rs]: Ok, it looks like all the remaining sysroot components are already built somewhere else and then copied to sysroot directory.

**Comment by petrochenkov** [comment]
Updated with suggestion from https://github.com/rust-lang/rust/pull/39892#discussion_r101650887
`rsbegin.o`/`rsend.o` are built in the `native` directrory (because they are shared between stages and because I didn't want to add another top-level directory) and then copied into sysroot directories.

**Comment by alexcrichton** [comment]
@bors: r+

Looks great!

**Pull request merged**